### ISSUE TITLE
Calling Dispatch in OnInitAsync in index.cshtml does not work

### ISSUE THREAD

**Issue by chanan** [mrpmorris/blazor-fluxor#28]
Calling Dispatcher.DispatchAsync from OnInitAsync in index.cshtml doesnt work. Either its broken in all pages now or just on the first page of the app?

```
MonoPlatform.ts:192 WASM: [System.InvalidOperationException] Store has not been initialized. Add `@Store.Initialize()` to your layout page
```

Note: Store is Initialized, and calling dispatch in onclick methods does work.

**Comment by mrpmorris** [comment]
Could you send a test case? This is done in the tutorials and is working.

**Comment by mrpmorris** [comment]
Are you still seeing this? If so then please send me a project that reproduces it.

**Comment by mrpmorris** [comment]
I'm going to close this. If you manage to create a repro then I will reopen it.

**Comment by vishalbiswas** [comment]
This seems to occur if you call Dispatch in OnInit of `@page "/"`

**Comment by vishalbiswas** [comment]
@mrpmorris I got this yesterday with the newest nuget package and blazor.
Will add exact versions later.

**Comment by mrpmorris** [comment]
Fixed in v0.15.0

**Comment by vishalbiswas** [comment]
Follow up on this, the Store.Initialize() bug seems to be solved in 0.15.

There appears to be another bug related to this. If you Dispatch in OnInit of @page "/" then the call is executed successfully but the injected IState object isn't updated in @page "/". The page inherits Blazor.Fluxor.Components.FluxorComponent.

**Comment by mrpmorris** [comment]
Can you provide an example app?

**Comment by vishalbiswas** [comment]
@mrpmorris Here -> https://github.com/vishalbiswas/TestBlazorFluxor

One thing to note, if I comment out `base.OnInit()` in Index.cshtml, then the button does work.

**Comment by mrpmorris** [comment]
@vishalbiswas It seems to be an issue with Blazor. I have recorded an issue on that project. I will keep this open until I receive a response.

https://github.com/aspnet/Blazor/issues/1400

**Comment by vishalbiswas** [comment]
That's what I was afraid of. Thanks for looking into it!

**Comment by mrpmorris** [comment]
@chanan @vishalbiswas Good news, I was able to work around this limitation by calling store.Initialize from a JavaScript setTimeOut(....., 0);

The fix is in version 0.15.1

**Comment by ransagy** [comment]
@mrpmorris This broke something in my use case that I'm not entirely sure is expected.
My `App` component inherits `FluxorComponent` and has a call to `Store.Initialize()`. The component itself holds some `IState<>` properties that come from the store and i use them during the `App.OnInitAsync()` life cycle method.

With v0.14, This was fine. In 0.15.1, The properties are now uninitialized/hold default values even after the store initialization. 

Am i doing something wrong? is there a new/better place to do the initializations with the new workaround?

**Comment by mrpmorris** [comment]
Hi

What does this code do that depends on the state already being initialised? If it is store related then it should be done in an that is triggered by StoreInitializedAction.

**Comment by mrpmorris** [comment]
Yes, that will fire only when the page initially loads and the store is initialised. This is where you should put code that you want to run once (and only once) when the store initialises regardless of which initial page you navigate to.

The sequence of events is as follows

1. Retrieve page from server
2. OnInitAsync is executed for any pages
3. Page is rendered in browser
4. JavaScript calls a page loaded event
5. Store is initialised
6. Any events dispatched up until now will now be dequeued

It's done this way because the page needs to have any middleware JavaScript inserted before the store starts to process anything, and that JavaScript can only be installed as the page loads. There might be another way to achieve what I need, but will the effect approach work for your current needs?

**Comment by ransagy** [comment]
Yeah, It likely will do. I don't know what's the best approach here for initializing the application, So `App.OnInit` seemed like a good place, But it might as well be part of the `StoreInitializedEffect`. Thanks!

**Comment by mrpmorris** [comment]
@ransagy Yes, try to do less in the view. The view is only there to present, initialisation of state should be done in the store.

**Comment by mrpmorris** [comment]
@chanan @vishalbiswas @ransagy 
I've just released version 0.16.0

If you need to ensure the store has initialized before dispatching (so you get immediate state updates) you can now do this

```
@inject IStore Store
functions {
  protected override async Task OnInitAsync() {
    await Store.Initialized;
    // Now dispatch
    // Now you can read updated state
  }
}
```

**Comment by ransagy** [comment]
Thanks, I'll give it a go. I worked around it for now by just passing the instance i initialized along inside `OnInit` but this would be cleaner.
